sunpy/util/scraper.py

```diff
@@ -0,0 +1,216 @@
+        # a list of all the blocks in times - assuming they are all
+        # separated with either '.', '_' or '/'
+        url_to_list = lambda txt: re.sub(r'\.|_', '/', txt).split('/')
+        pattern_list = url_to_list(self.pattern)
+        url_list = url_to_list(url)
+
+        time_order = ['%Y', '%y', '%b', '%B', '%m', '%d', '%j',
+                      '%H', '%I', '%M', '%S']
+        final_date = []
+        final_pattern = []
+        # Find in directory and filename
+        for pattern_elem, url_elem in zip(pattern_list, url_list):
+            time_formats = [x for x in time_order if x in pattern_elem]
+            if len(time_formats) > 0:
+                final_date.append(url_elem)
+                final_pattern.append(pattern_elem)
+                for time_bit in time_formats:
+                    time_order.remove(time_bit)
+        # Find and remove repeated elements eg: %Y in ['%Y', '%Y%m%d']
+        #   Make all as single strings
+        date_together = ''.join(final_date)
+        pattern_together = ''.join(final_pattern)
+        re_together = pattern_together
+        for k, v in TIME_CONVERSIONS.iteritems():
+            re_together = re_together.replace(k, v)
+
+        #   Create new empty lists 
+        final_date = list()
+        final_pattern = list()
+        for p,r in zip(pattern_together.split('%')[1:], re_together.split('\\')[1:]):
+            regexp = '\\{}'.format(r)
+            pattern = '%{}'.format(p)
+            date_part = re.match(regexp, date_together)
+            date_together = date_together[:date_part.start()] + \
+                            date_together[date_part.end():]
+            if pattern not in final_pattern:
+                final_pattern.append('%{}'.format(p))
+                final_date.append(date_part.group())
+        return datetime.datetime.strptime(' '.join(final_date),
+                                          ' '.join(final_pattern))
+
+    def filelist(self, timerange):
+        """
+        Returns the list of existent files in the archive for the
+        given time range.
+
+        Parameters
+        ----------
+
+        timerange : `~sunpy.time.TimeRange`
+            Time interval where to find the directories for a given
+            pattern.
+
+        Returns
+        -------
+
+        filesurls : list of strings
+            List of all the files found between the time range given.
+
+        Examples
+        --------
+        >>> from sunpy.time import TimeRange
+        >>> timerange = TimeRange('2015-01-01','2015-01-01T16:00:00')
+        >>> print(solmon.filelist(timerange))
+        ['http://solarmonitor.org/data/2015/01/01/fits/swap/swap_00174_fd_20150101_025423.fts.gz']
+        """
+        directories = self.range(timerange)
+        filesurls = []
+        for directory in directories:
+            try:
+                opn = urllib2.urlopen(directory)
+                try:
+                    soup = BeautifulSoup(opn)
```

Comment:
humm... that doesn't seem to work.  Or at least not by just changing that :(

I get an 
```python
AttributeError: __exit__
```

I've also tried to change it in the line above:
```python
with urllib2.urlopen(directory) as opn:
    ...
```
but also I get an 
```python
AttributeError: addinfourl instance has no attribute '__exit__'
```